### PR TITLE
[refactoring] Relax trait bounds, reduce dependence on `StateView`

### PR DESCRIPTION
**Pull request by georgemitenkov**
### Description

API only needs move resolver, and not state view. Decoupling
`MoveResolverExt` into `AptosMoveResolver + StateView` so
that any future changes were not visible on API side.

1. Moved API checks for resource groups where it should be.

2. Making trait bounds for config not require state view. Right now
state view is used all over the place although it seems this can
be avoided.

3. Reducing the usage of `MoveResolverExt`: e.g. session only
needs to talk to resolver.

4. State value metadata has its own resolver. This makes storage
accesses almost uniform (still need to fix `read_write_set`) which
should be implemented in storage adapter.

5. Some minor typo fixes and dead code removal.

@msmouse @davidiw what do you think?

### CODE REVIEW

File: .github/linters/semgrep/pull-request-target-code-checkout.yaml [hunk 1]
```diff
@@ -47,6 +47,15 @@ rules:
             ...
             $JOBNAME:
               ...
+      - pattern-not-inside: |
+          needs: [..., permission-check, ...]
```
Comment (vusirikala) [inline .github/linters/semgrep/pull-request-target-code-checkout.yaml:51]: What are these changes? These changes seem to be in the linters directory.

Comment (georgemitenkov) [reply .github/linters/semgrep/pull-request-target-code-checkout.yaml:51]: I think it is a CI fix, will rebase later

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 2]
```diff
@@ -235,3 +233,30 @@ impl<S: StateView> AsMoveResolver<S> for S {
...
     }
 }
+
+impl<'a, S: StateView> StateValueMetadataResolver for StorageAdapter<'a, S> {
+    fn get_state_value_metadata(
+        &self,
+        state_key: &StateKey,
+    ) -> anyhow::Result<Option<Option<StateValueMetadata>>> {
+        let maybe_state_value = self.state_store.get_state_value(state_key)?;
+        Ok(maybe_state_value.map(|state_value| state_value.into_metadata()))
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:243]: So the outer Option is because of the maybe_state_value being empty, and inside is if the into_metadata doesn't succeed? can't we map_err to storage error on the outer one here to convert it to the result? (seems to be the way the caller uses it)?

**Review by gelash — APPROVED** [review]

File: aptos-move/aptos-vm/src/move_vm_ext/resolver.rs [hunk 3]
```diff
@@ -30,18 +49,9 @@ pub trait MoveResolverExt:
...
-                .iter()
-                .find(|attr| attr.is_resource_group())?;
-            Some(())
-        })
-        .is_some()
-    }
 }
+
+// TODO: Remove dependency on StateView.
+pub trait MoveResolverExt: AptosMoveResolver + StateView {}
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/resolver.rs:55]: We need it for `StateView` in `RespawnedSession` (so executing user transaction essentially) and `read_write_set()`.

**Review by zekun000 — APPROVED** [review]
the respawned session is a bit ugly but probably for another pr

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 4]
```diff
@@ -119,16 +118,21 @@ macro_rules! unwrap_or_discard {
 }
 
 impl AptosVM {
-    pub fn new(state: &impl StateView) -> Self {
-        Self(AptosVMImpl::new(state))
+    pub fn new(config_storage: &impl ConfigStorage) -> Self {
```
Comment (vusirikala) [inline aptos-move/aptos-vm/src/aptos_vm.rs:121]: What is `ConfigStorage`? How is it different from `StateView`? Why is `ConfigStorage` enough here instead of `StateView`?

File: api/src/context.rs [hunk 5]
```diff
@@ -43,9 +43,10 @@ use aptos_types::{
     },
     transaction::{SignedTransaction, TransactionWithProof, Version},
 };
+use aptos_utils::aptos_try;
 use aptos_vm::{
     data_cache::{AsMoveResolver, StorageAdapter},
-    move_vm_ext::MoveResolverExt,
+    move_vm_ext::AptosMoveResolver,
```
Comment (vusirikala) [inline api/src/context.rs:49]: Is `AptosMoveResolver` just a renaming for `MoveResolverExt`? Or are they different in functionality?

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 6]
```diff
@@ -48,7 +52,7 @@ pub struct StorageAdapter<'a, S> {
         RefCell<BTreeMap<AccountAddress, BTreeMap<StructTag, BTreeMap<StructTag, Vec<u8>>>>>,
 }
 
-impl<'a, S: StateView> StorageAdapter<'a, S> {
+impl<'a, S> StorageAdapter<'a, S> {
```
Comment (vusirikala) [inline aptos-move/aptos-vm/src/data_cache.rs:55]: Why is `StateView` trait bound removed here? Isn't StorageAdapter defined only for StateView?

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 7]
```diff
@@ -235,3 +233,30 @@ impl<S: StateView> AsMoveResolver<S> for S {
...
+//      in epilogue.
+impl<'a, S: StateView> TStateView for StorageAdapter<'a, S> {
+    type Key = StateKey;
+
+    fn get_state_value(&self, state_key: &Self::Key) -> anyhow::Result<Option<StateValue>> {
+        self.state_store.get_state_value(state_key)
+    }
+
+    fn get_usage(&self) -> anyhow::Result<StateStorageUsage> {
+        self.state_store.get_usage()
```
Comment (vusirikala) [inline aptos-move/aptos-vm/src/data_cache.rs:260]: What is usage here?

File: aptos-move/aptos-vm/src/move_vm_ext/session.rs [hunk 8]
```diff
@@ -428,14 +428,17 @@ impl<'r> WriteOpConverter<'r> {
...
 
-        let existing_value_opt = self.remote.get_state_value(state_key).map_err(|_| {
-            VMStatus::error(
-                StatusCode::STORAGE_ERROR,
-                err_msg("Storage read failed when converting change set."),
-            )
-        })?;
-
-        let write_op = match (existing_value_opt, move_storage_op) {
+        let maybe_existing_metadata =
```
Comment (vusirikala) [inline aptos-move/aptos-vm/src/move_vm_ext/session.rs:431]: Instead of using existing value, we are now using the metadata of existing value. What's the motivation behind this change?

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 4 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/aptos_vm.rs:121]: `ConfigStorage` is an abstraction which allows you to get resources which are configs, e.g. gas schedule, etc. In case of VM, we don't care about state view when we initialise, all we care is the configs.

File: aptos-move/aptos-vm/src/data_cache.rs (hunk 6 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/data_cache.rs:55]: There are 2 impls: all functionality relies on state view, but one should be able to create an adapter on top of some other type and implement move resolver / state view on top.

File: aptos-move/aptos-vm/src/data_cache.rs (hunk 7 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/data_cache.rs:260]: I think this is needed for Move functions to query how many bytes are used in storage, etc. for pricing.

File: aptos-move/aptos-vm/src/move_vm_ext/session.rs (hunk 8 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/session.rs:431]: We never needed the value there - only metadata to calculate storage refunds. Motivation behind having a resolver is to avoid using a fixed `get_state_value` method and use resolver abstraction so we can use any storage backend - for indexer, execution, etc.

Maybe we should expose this in resource/move resolvers, but this suffices for now I think

File: aptos-move/aptos-vm/src/data_cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/data_cache.rs:243]: Not really: the outer thing can be None if the state value does not exist: in this case for storage refunds this is acceptable. If inner does not exist, then there is no metadata associated, so we kind of need both...

I guess using named enums instead of option would be nicer in general😄

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs [hunk 9]
```diff
@@ -30,7 +31,10 @@ impl<'a, S: 'a + StateView + Sync> ExecutorTask for AptosExecutorTask<'a, S> {
     type Txn = PreprocessedTransaction;
 
     fn init(argument: &'a S) -> Self {
-        let vm = AptosVM::new(argument);
+        // AptosVM has to be initialized using configs from storage.
+        // Using adapter allows us to fetch those.
+        let config_storage = StorageAdapter::new(argument);
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:36]: Oh, so for this change doing this allows us to not have a state view bound here actually, so it's easier to implement things in the future (I started doing that and ended up refactoring). Maybe using `from_state_view` which hides it is better?

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 10]
```diff
@@ -1231,7 +1235,7 @@ impl AptosVM {
         }
     }
 
-    fn read_writeset<'a>(
+    fn read_write_set<'a>(
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/aptos_vm.rs:1238]: renamed to `read_change_set` instead, should be cleaner (and this takes change set as argument now)

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:36]: Added a comment about that

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 11]
```diff
@@ -235,3 +233,30 @@ impl<S: StateView> AsMoveResolver<S> for S {
...
+        Ok(maybe_state_value.map(|state_value| state_value.into_metadata()))
+    }
+}
+
+// We need to implement StateView for adapter because:
+//   1. When processing write set payload, storage is accessed
+//      directly.
+//   3. When stacking Storage adapters on top of each other, e.g.
+//      in epilogue.
+impl<'a, S: StateView> TStateView for StorageAdapter<'a, S> {
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/data_cache.rs:252]: Nice! But this crate does not allow trait implementations, only function call resolution? And here we need StorageAdapter to satisfy trait bounds. Not sure how it helps?

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:252]: lol, for that there also seems to be a crate: https://docs.rs/ambassador/latest/ambassador/
but I'm not sure using two crates to save 5 lines is worth it )

[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/data_cache.rs:252]: Yep, I think with all the impls delegate can be pretty useful actually :) not sure about here, can revisit this, and anyway with new traits we will not need StateView here
